x/ibc/07-tendermint/types/proposal_handle_test.go

```diff
@@ -0,0 +1,256 @@
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = false
+			}, false, false,
+		},
+
+		{
+			"allowed to be updated only after expiry, client is frozen", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = false
+
+				clientState.FrozenHeight = frozenHeight
+			}, false, false,
+		},
+		{
+			"PASS: allowed to be updated only after expiry, client is expired", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = false
+
+				suite.chainA.ExpireClient(clientState.TrustingPeriod)
+				suite.chainB.ExpireClient(clientState.TrustingPeriod)
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+			}, true, true, // unfreezing headers work since they pass stricter checks
+		},
+
+		{
+			"allowed to be updated only after expiry, client is frozen and expired", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = false
+
+				clientState.FrozenHeight = frozenHeight
+				suite.chainA.ExpireClient(clientState.TrustingPeriod)
+				suite.chainB.ExpireClient(clientState.TrustingPeriod)
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+			}, false, false,
+		},
+		{
+			"allowed to be updated after expiry and misbehaviour, not frozen or expired", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = true
+			}, false, false,
+		},
+		{
+			"PASS: allowed to be updated after expiry and misbehaviour, client is frozen", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = true
+
+				clientState.FrozenHeight = frozenHeight
+			}, true, false,
+		},
+		{
+			"PASS: allowed to be updated after expiry and misbehaviour, client is expired", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = true
+
+				suite.chainA.ExpireClient(clientState.TrustingPeriod)
+				suite.chainB.ExpireClient(clientState.TrustingPeriod)
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+			}, true, true, // unfreezing headers work since they pass stricter checks
+		},
+
+		{
+			"PASS: allowed to be updated after expiry and misbehaviour, client is frozen and expired", func() {
+				clientState.AllowUpdateAfterExpiry = true
+				clientState.AllowUpdateAfterMisbehaviour = true
+
+				clientState.FrozenHeight = frozenHeight
+				suite.chainA.ExpireClient(clientState.TrustingPeriod)
+				suite.chainB.ExpireClient(clientState.TrustingPeriod)
+				suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+			}, true, true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		// for each test case a header used for unexpiring clients and unfreezing
+		// a client are each tested to ensure the work as expected and that
+		// unexpiring headers cannot unfreeze clients.
+		suite.Run(tc.name, func() {
```

Comment:
Here with each test case construct the clientState with the provided override flags.

1. Test UpdateProposal against a valid clientstate (this is optional, you can remove this step)
2. Freeze the client and provide the update header and see if you get the expected result
3. Unfreeze and expire the client and provide the update header and see if you get the expected result
4. Freeze and expire the client and provide the update header and see if you get the expected result

This would be much clearer imo, and require less test-case structs

Reply:
I don't understand this reduction. Each of the test cases in one of the four categories have varying results for the combo of `expPassUnfreeze` and `expPassUnexpire`? Trying to program this logic into the `suite.Run` becomes messy and harder to reason about imo. I don't think I can cleanly reduce the test cases to 4 without dropping some testing logic (not testing unexpiry headers on frozen clients etc). I think it is important to test weaker headers on test cases expected to have stronger headers to ensure they don't pass.

I can abstract the clientState to be constructed on the override flags. I think any other improvement can be done in a followup (or you can push directly) because I don't follow.

Reply:
I can also do valid client state test initially to reduce the test case # by 4